api/validator.go

```diff
@@ -0,0 +1,143 @@
+					return fmt.Errorf("spec.rule[%d].http.nodePort %s may not be specified when `LBType` is `HostPort`", ri, rule.HTTP.NodePort)
+				}
+				if ei, found := nodePorts[np]; found {
+					return fmt.Errorf("spec.rule[%d].http is reusing nodePort %s for addr %s, also used in spec.rule[%d]", ri, np, addr, ei)
+				} else {
+					nodePorts[np] = ri
+				}
+			}
+
+			paths := make(map[string]int)
+			for pi, path := range rule.HTTP.Paths {
+				if ei, found := paths[path.Path]; found {
+					return fmt.Errorf("spec.rule[%d].http.paths[%d] is reusing path %s for addr %s, also used in spec.rule[%d].http.paths[%d]", ri, pi, path, addr, ri, ei)
+				}
+				paths[path.Path] = pi
+
+				if path.Backend.ServiceName == "" {
+					return fmt.Errorf("spec.rule[%d].http.paths[%d] is missing serviceName for addr %s and path %s", ri, pi, addr, path.Path)
+				}
+				if _, err := checkRequiredPort(path.Backend.ServicePort); err != nil {
+					return fmt.Errorf("spec.rule[%d].http.paths[%d] is using invalid servicePort %s for addr %s and path %s. Reason: %s", ri, pi, path.Backend.ServicePort, addr, path.Path, err)
+				}
+
+				for hi, hdr := range path.Backend.HeaderRule {
+					if len(strings.Fields(hdr)) == 1 {
+						return fmt.Errorf("spec.rule[%d].http.paths[%d].backend.headerRule[%d] is invalid for addr %s and path %s.", ri, pi, hi, addr, path.Path)
+					}
+				}
+
+			}
+		} else if rule.TCP != nil {
+			addr := address{Host: rule.Host}
+			if port, err := checkRequiredPort(rule.TCP.Port); err != nil {
+				return fmt.Errorf("spec.rule[%d].tcp.port %s is invalid. Reason: %s", ri, rule.TCP.Port, err)
+			} else {
+				addr.Port = port
+			}
+			if ei, found := addrs[addr]; found {
+				return fmt.Errorf("spec.rule[%d].tcp is reusing addr %s, also used in spec.rule[%d]", ri, addr, ei)
+			}
+			addrs[addr] = ri
+
+			if np, err := checkOptionalPort(rule.TCP.NodePort); err != nil {
+				return fmt.Errorf("spec.rule[%d].tcp.nodePort %s is invalid. Reason: %s", ri, rule.TCP.NodePort, err)
+			} else if np > 0 {
+				if r.LBType() == LBTypeHostPort {
+					return fmt.Errorf("spec.rule[%d].tcp.nodePort %s may not be specified when `LBType` is `HostPort`", ri, rule.TCP.NodePort)
+				}
+				if ei, found := nodePorts[np]; found {
+					return fmt.Errorf("spec.rule[%d].tcp is reusing nodePort %s for addr %s, also used in spec.rule[%d]", ri, np, addr, ei)
+				} else {
+					nodePorts[np] = ri
+				}
+			}
+
+			if rule.TCP.Backend.ServiceName == "" {
+				return fmt.Errorf("spec.rule[%d].tcp is missing serviceName for addr %s", ri, addr)
+			}
+			if _, err := checkRequiredPort(rule.TCP.Backend.ServicePort); err != nil {
+				return fmt.Errorf("spec.rule[%d].tcp is using invalid servicePort %s for addr %s. Reason: %s", ri, rule.TCP.Backend.ServicePort, addr, err)
+			}
+			if len(rule.TCP.Backend.HeaderRule) > 0 {
+				return fmt.Errorf("spec.rule[%d].tcp.backend.headerRule must be empty for addr %s", ri, addr)
+			}
+			if len(rule.TCP.Backend.RewriteRule) > 0 {
```

Comment:
We can do this. But user can still do whatever they want in YAML. I believe the marshaler just ignores unknown fields.